src/riscv/lib/src/machine_state.rs

```diff
@@ -580,6 +633,64 @@ impl<MC: memory::MemoryConfig, BCC: BlockCacheConfig, B: Block<MC, M>, M: backen
     }
 }
 
+/// The alignment of a stack pointer in RISC-V's ABI
+/// See RISC-V ABIs Specification Chapter 2.1
+pub const RISCV_ABI_SP_ALIGNMENT: u64 = 16;
+
+/// Can be pushed to a stack frame.
+/// Ensures that elements are pushed onto a stack frame in the correct order.
+///
+/// See also [StackFramePop]
+pub trait StackFramePush: Default + Sized {
+    type Above: StackFramePop;
+
+    #[must_use = "The stack frame must be initialised in full"]
+    fn push<MC: MemoryConfig, BCC: BlockCacheConfig, B: Block<MC, M>, M: ManagerReadWrite>(
+        &self,
+        state: &mut MachineState<MC, BCC, B, M>,
+        data: <Self::Above as StackFramePop>::Data,
+    ) -> Result<Self::Above, MachineError> {
+        if <Self::Above as StackFramePop>::ALIGN {
+            state.push_stack::<_, true>(std::slice::from_ref(&data))?;
+        } else {
+            state.push_stack::<_, false>(std::slice::from_ref(&data))?;
+        }
+        Ok(Self::Above::pushed::<Self>())
+    }
+
+    fn popped<Above>() -> Self {
+        Default::default()
+    }
+}
+
+/// Can be popped from a stack frame.
+/// Ensures that elements are popped from a stack frame in the correct order.
+///
+/// See also [StackFramePush]
+pub trait StackFramePop: Default + Sized {
+    type Data: Elem + Copy;
+    type Below: StackFramePush;
+    const ALIGN: bool = true;
+
+    #[must_use = "The stack frame must be restored after use"]
+    fn pop<MC: MemoryConfig, BCC: BlockCacheConfig, B: Block<MC, M>, M: ManagerReadWrite>(
+        &self,
+        state: &mut MachineState<MC, BCC, B, M>,
+        data: &mut Self::Data,
+    ) -> Result<Self::Below, MachineError> {
+        if Self::ALIGN {
```

Comment:
If you make `ALIGN` a bool const generic on `StackFramePop/Push` you can get rid of the if statement